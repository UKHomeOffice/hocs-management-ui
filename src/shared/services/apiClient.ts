import axios from 'axios';

export default {
    createClient: () => {
        const client = axios.create();

        return {
            get: client.get,
            delete: client.delete,
            post: client.post,
            put: client.put
        };
    }
};
