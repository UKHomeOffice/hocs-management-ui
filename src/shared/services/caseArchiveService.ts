import axios from 'axios';
import ArchiveCaseModel from '../models/archiveCaseModel';

export const archiveCase = (archiveCaseModel: ArchiveCaseModel) => new Promise((resolve, reject) => axios
    .post('/api/case/archive', archiveCaseModel)
    .then(() => resolve(null))
    .catch(reason => reject(reason))
);
