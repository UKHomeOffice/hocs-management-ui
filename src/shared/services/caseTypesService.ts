import axios from 'axios';
import CaseType from '../models/caseType';
import Item from '../models/item';

const sortByLabel: ((a: Item, b: Item) => number) = (a, b) => {
    const labelA = a.label.toLocaleLowerCase();
    const labelB = b.label.toLocaleLowerCase();
    let comparison = 0;
    if (labelA > labelB) {
        comparison = 1;
    } else if (labelA < labelB) {
        comparison = -1;
    }
    return comparison;
};

export const getCaseTypes = () => new Promise<CaseType[]>((resolve, reject) =>
    axios.get('/api/case-types')
        .then((response) => {
            resolve(response.data.sort(sortByLabel));
        })
        .catch(error => reject(error))
);
