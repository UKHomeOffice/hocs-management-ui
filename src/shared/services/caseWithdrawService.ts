import axios from 'axios';
import WithdrawCaseModel from '../models/withdrawCaseModel';

export const withdrawCase = (withdrawCaseModel: WithdrawCaseModel) => new Promise((resolve, reject) => axios
    .post('/api/case/withdraw', withdrawCaseModel)
    .then(() => resolve(null))
    .catch(reason => reject(reason))
);
