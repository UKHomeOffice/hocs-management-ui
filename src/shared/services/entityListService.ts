import axios from 'axios';
import EntityListItem from 'shared/models/entityListItem';

export const getListItems = (listName: string) => new Promise<EntityListItem[]>((resolve, reject) => axios
    .get(`/api/entity/list/${listName}`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const createListItem = (item: EntityListItem, listName: string) => new Promise((resolve, reject) => axios
    .post(`/api/entity/list/${listName}`, item)
    .then(() => resolve(null))
    .catch(reason => reject(reason))
);

export const getItemDetails = (itemUUID: string) => new Promise<EntityListItem>((resolve, reject) => axios
    .get(`/api/entity/${itemUUID}`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const updateListItem = (item: EntityListItem, listName: string) => new Promise((resolve, reject) => axios
    .post(`/api/entity/list/update/${listName}`, item)
    .then(() => resolve(null))
    .catch(reason => reject(reason))
);

export const deleteListItem = (uuid: string, listName: string) => new Promise((resolve, reject) => axios
    .delete(`/api/entity/list/${listName}`, { data: { entityUUID: uuid } })
    .then(() => resolve(null))
    .catch(reason => reject(reason))
);
