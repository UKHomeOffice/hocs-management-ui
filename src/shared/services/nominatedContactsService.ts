import axios from 'axios';
import NominatedContact from '../models/nominatedContact';
import Contact from '../models/contact';

export const addNominatedContact = (nominatedContact: NominatedContact) => new Promise((resolve, reject) => axios
    .post('/api/nominated-contact', nominatedContact)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const getNominatedContactsForTeam = (teamUUID: string) => new Promise<Array<Contact>>((resolve, reject) => axios
    .get(`/api/nominated-contact/${teamUUID}/`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const removeNominatedContactFromTeam = (teamUUID: string, nominatedContactUUID: string, ) => new Promise((resolve, reject) => axios
    .delete(`/api/nominated-contact/${teamUUID}/${nominatedContactUUID}`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);
