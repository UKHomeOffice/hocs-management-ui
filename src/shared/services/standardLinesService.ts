import axios from 'axios';
import UpdateRequest from 'shared/pages/standardLineAmend/updateRequest';

export const addStandardLine = (standardLine: FormData) => new Promise((resolve, reject) => axios
    .post('/api/standard-lines', standardLine)
    .then(() => resolve())
    .catch(reason => reject(reason))
);

export const updateStandardLine = (standardLine: UpdateRequest) => new Promise((resolve, reject) => axios
    .post('/api/standard-lines/update', standardLine)
    .then(() => resolve())
    .catch(reason => reject(reason))
);

export const getAllStandardLines = () => new Promise((resolve, reject) => axios
    .get('/api/standard-lines/all')
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const getStandardLine = (standardLineUuid: string) => new Promise((resolve, reject) => axios
    .get(`/api/standard-lines/get/${standardLineUuid}`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const expireStandardLine = (standardLineUuid: string) => new Promise((resolve, reject) => axios
    .post(`/api/standard-lines/expire/${standardLineUuid}`, {})
    .then(() => resolve())
    .catch(reason => reject(reason))
);

export const deleteStandardLine = (standardLineUuid: string) => new Promise((resolve, reject) => axios
    .delete(`/api/standard-lines/delete/${standardLineUuid}`, {})
    .then(() => resolve())
    .catch(reason => reject(reason))
);

export const downloadStandardLine = (standardLineUuid: string) => new Promise((resolve, reject) => axios
    .get(`/api/standard-lines/download/${standardLineUuid}`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);
