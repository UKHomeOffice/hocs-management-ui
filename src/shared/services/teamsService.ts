import axios from 'axios';
import Item from '../models/item';
import Team from '../models/team';

export const getTeam = (teamId: string) => new Promise<Team>((resolve, reject) =>
    axios.get(`/api/teams/${teamId}`)
        .then(response => resolve(response.data))
        .catch(error => reject(error))
);

export const getTeams = () => new Promise<[Item]>((resolve, reject) =>
    axios.get('/api/teams')
        .then(response => resolve(response.data))
        .catch(error => reject(error))
);

export const getTeamMembers = (teamId: string) => new Promise((resolve, reject) => axios
    .get(`/api/teams/${teamId}/members`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const getTeamsForUser = (userId: string) => new Promise<[Item]>((resolve, reject) => axios
    .get(`/api/teams/${userId}/teams`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const createTeam = (team: Team) => new Promise((resolve, reject) => axios
    .post(`/api/teams/unit/${team.unitUUID}`, team)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);
