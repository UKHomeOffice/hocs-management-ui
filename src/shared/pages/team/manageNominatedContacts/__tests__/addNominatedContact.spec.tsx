import React from 'react';
import { act, fireEvent, getByText, render, RenderResult, wait, getByLabelText } from '@testing-library/react';
import AddNominatedContact from '../addNominatedContacts';
import * as NominatedContactsService from '../../../../services/nominatedContactsService';

const mockAddFormError = jest.fn();
const mockClearErrors = jest.fn();
const mockSetErrorMessage = jest.fn();
const mockDispatch = jest.fn();

jest.spyOn(React, 'useContext').mockImplementation(() => ({
    dispatch: mockDispatch
}));

const renderComponent = () => render(
    <AddNominatedContact teamId="testTeam" errorFuncs={[null, mockAddFormError, mockClearErrors, mockSetErrorMessage]}/>
);

const addNominatedContactSpy = jest.spyOn(NominatedContactsService, 'addNominatedContact');

describe('when the AddNominatedContact component is mounted', () => {
    it('should render with default props', async () => {
        let wrapper: RenderResult;
        act(() => {
            wrapper = renderComponent();
        });
        expect.assertions(1);

        await wait(() => {
            expect(wrapper.container).toMatchSnapshot();
        });
    });
});

describe('when the submit button is clicked', () => {
    let wrapper: RenderResult;

    beforeEach(() => {
        mockAddFormError.mockReset();
        mockClearErrors.mockReset();
        mockSetErrorMessage.mockReset();
        mockDispatch.mockReset();
        addNominatedContactSpy.mockReset();

        act(() => {
            wrapper = renderComponent();
        });
    });

    const updateEmailAddressInputText = async (email: string) => {
        act(() => {
            const emailAddressInput = getByLabelText(wrapper.container, 'Email Address');
            fireEvent.change(emailAddressInput, { target: { value: email } });
        });
    };

    const fireSubmit = async () => {
        act(() => {
            const submitButton = getByText(wrapper.container, 'Add nominated contact');
            fireEvent.click(submitButton);
        });
    };


    it('should show errors and not attempt to submit address which fail validation', async () => {
        updateEmailAddressInputText('');
        fireSubmit();

        await wait(async () => {
            expect(mockClearErrors).toBeCalledTimes(1);
            expect(addNominatedContactSpy).toBeCalledTimes(0);
            expect(mockDispatch).toBeCalledTimes(0);

            expect(mockAddFormError).nthCalledWith(1, {
                'key': 'inputValue',
                'value': 'The email address is required',
            });

            expect(mockAddFormError).nthCalledWith(2, {
                'key': 'inputValue',
                'value': 'The email address contains invalid characters'
            });
        });
    });

    it('should show the correct error for duplicate email', async () => {

        addNominatedContactSpy.mockImplementationOnce(() => Promise.reject({
            response: {
                status: 409
            }
        }));

        updateEmailAddressInputText('a@a.example.org');
        fireSubmit();

        await wait(async () => {
            expect(mockClearErrors).toBeCalledTimes(1);
            expect(addNominatedContactSpy).toBeCalledTimes(1);
            expect(mockDispatch).toBeCalledTimes(0);

            expect(mockSetErrorMessage).toBeCalledWith({
                'description': 'A nominated contact with those email details already exists',
                'title': 'There was a error validating the response'
            });
        });
    });

    it('should show a generic error for errors other than duplicate email', async () => {
        addNominatedContactSpy.mockImplementationOnce(() => Promise.reject({
            response: {
                status: 400
            }
        }));

        updateEmailAddressInputText('a@a.example.org');
        fireSubmit();

        await wait(async () => {
            expect(mockClearErrors).toBeCalledTimes(1);
            expect(addNominatedContactSpy).toBeCalledTimes(1);
            expect(mockDispatch).toBeCalledTimes(0);

            expect(mockSetErrorMessage).toBeCalledWith({
                'description': 'Something went wrong while adding the nominated contact. Please try again.',
                'title': 'Something went wrong'
            });
        });
    });

    it('should call the service and update the context on success', async () => {

        addNominatedContactSpy.mockImplementationOnce(() => Promise.resolve({ uuid: 'new-uuid' }));

        updateEmailAddressInputText('abc@example.org');
        fireSubmit();

        await wait(async () => {
            expect(mockClearErrors).toBeCalledTimes(1);

            expect(addNominatedContactSpy).toBeCalledWith({
                emailAddress: 'abc@example.org',
                teamUUID: 'testTeam'
            });

            expect(mockDispatch).toBeCalledWith({
                'payload': {
                    'label': 'abc@example.org',
                    'value': 'new-uuid',
                },
                'type': 'AddContact'
            });
        });
    });
});
