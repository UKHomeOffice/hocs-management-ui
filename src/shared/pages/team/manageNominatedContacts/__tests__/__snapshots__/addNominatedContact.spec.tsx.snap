// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the AddNominatedContact component is mounted should render with default props 1`] = `
<div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-one-half-from-desktop"
    >
      <form
        method="POST"
      >
        <input
          name="_csrf"
          type="hidden"
          value=""
        />
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="emailAddress"
            id="emailAddress-label"
          >
            Email Address
          </label>
          <input
            class="govuk-input"
            id="emailAddress"
            name="emailAddress"
            type="text"
            value=""
          />
        </div>
        <input
          class="govuk-button"
          type="submit"
          value="Add nominated contact"
        />
      </form>
    </div>
  </div>
</div>
`;
